packages/documentation/copy/en/reference/Target Language Details.md

```diff
@@ -2673,6 +2673,23 @@ Note that when building with a release profile (i.e., target property `build-typ
 
 </div>
 
+<div class="lf-c">
+
+## Scheduler Target Property
+
+The `scheduler` target property is used to select the scheduler used by the C runtime. This scheduler determines the order in which reactions are processed. It also assigns reactions to user-level threads and can thereby influence the assignment of reactions to processors.
+
+Because the C runtime scheduler operates at a higher level of abstraction than the OS, none of the scheduling policies that we currently support allow preemption; furthermore, they do not control migration of threads between processors.
+
+The following schedulers are available:
+
+- `GEDF_NP` (global earliest-deadline-first): This scheduler is the default scheduler for programs that have deadlines. It guarantees that in the execution of any given tag, reactions with earlier deadlines are executed before reactions with later deadlines; however, the tag of a reaction will still take precedence over its deadline in determining execution order. Reactions with no explicit deadline implicitly have an infinitely late deadline.
```

Comment:
> It guarantees that in the execution of any given tag, reactions with earlier deadlines are executed before reactions with later deadlines

I think this is technically true, but it is glossing over the fact that reaction precedence within the same tag is also important. E.g., the order of execution for the following two reactions is not according to the user-provided deadline:
```
main reactor {
    reaction(startup) {= =} deadline(2 msec) {= =}
    reaction(startup) {= =} deadline(1 msec) {= =}
}
```

Reply:
You put this delicately @Soroosh129; what I wrote was pretty wrong. I just rewrote and pushed my changes, but the description is so obscure and implementation-specific that I wonder if we would prefer to simply describe it as a best-effort policy that respects deadlines. What do you think?

Reply:
How about this:
> `GEDF_NP` (global earliest-deadline-first): This scheduler is the default scheduler for programs that have deadlines. Whenever the semantics of Lingua Franca allows the concurrent execution of two or more ready reactions at a particular tag, then this scheduler will prioritize the reaction with the earliest deadline to run first (but with the limitation that reaction executions are non-preemptive). Reactions with no explicit deadline implicitly have an infinitely late deadline.

Reply:
Well, that is what I wanted to say, but unless I am mistaken, the scheduler does use levels. It is possible for a reaction be have all its dependencies satisfied and to have an earlier deadline than reactions in the currently executing level, but to have its execution deferred because we process the reaction graph in level order.

Reply:
Levels are just one way of encoding reaction precedence (another way would be via a topologically sorted directed graph). I tried to abstract this detail away and simply say "Whenever the semantics of Lingua Franca allows for concurrent execution of two or more ready reactions at a particular tag". In the C runtime, that means that whenever two reactions are ready (and triggered) at the current tag and have the same level. I think the only substantial thing I added to your previous explanation was the aforementioned sentence.

Reply:
But... levels are not part of the semantics of Lingua Franca. If they were, then we would not be able to use chain ID or relaxation of the barrier synchronization (LET).

This is really just a detail, though. I almost merged something that was even less precise.

Reply:
I see your point. The `GEDF_NP` scheduler's qualification for what reactions can execute concurrently is very limited and is only a subset of what LF semantics would allow. 

Then, perhaps, levels are worth explaining briefly as you did. One note that I have is that all the currently-listed schedulers use levels in some form. Perhaps levels merit a brief sub-subsection here?